Move to Dockerfile installation

tyler36 opened the pull request:
Fixes #31

tyler36 committed:
Move to Dockerfile installation

tyler36 committed:
fix README typos

tyler36 committed:
Update readme

rfay commented on README.md:
````diff
@@ -2,77 +2,80 @@
 
-## Implementation
+  ```shell
+  ddev get ddev/ddev-browsersync
````
OOps, wrong addon!

```suggestion
  ddev get ddev/ddev-cron
```

rfay commented on README.md:
```diff
@@ -2,77 +2,80 @@
-The provided `web-build/Dockerfile.ddev-cron` and `web-build/cron.conf` configure the traditional cron daemon to run inside the web container.
+- Update `./.ddev/web-build/custom.cron` with your required commands.
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
+- Custom the cron service, if required, by updating `./.ddev/web-build/cron.conf`
```
```suggestion
- Customize the cron service, if required, by updating `./.ddev/web-build/cron.conf`
```

rfay commented on README.md:
```diff
@@ -2,77 +2,80 @@
 
-The provided `web-build/Dockerfile.ddev-cron` and `web-build/cron.conf` configure the traditional cron daemon to run inside the web container.
+- Update `./.ddev/web-build/custom.cron` with your required commands.
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
```
I think it might be better to try something that didn't require editing the #ddev-generated file, like maybe telling them to add another cron file that is *not* managed?

tyler36 replied:
This section has no longer exists.

rfay commented on README.md:
```diff
@@ -2,77 +2,80 @@
+- Update `./.ddev/web-build/custom.cron` with your required commands.
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
+- Custom the cron service, if required, by updating `./.ddev/web-build/cron.conf`
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
```
Again, if we can use override files instead of editing the installed files it will be better. We want people to always be able to `ddev get ddev/ddev-cron` without having trouble with their changes.

tyler36 replied:
I wonder if it's OK to keep the `cron.conf` file as `#ddev-generated` as this is what we do with nginx, postgres. The current defaults seem sensible for different projects.

rfay commented on README.md:
```diff
@@ -2,77 +2,80 @@
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
+- Custom the cron service, if required, by updating `./.ddev/web-build/cron.conf`
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
+- Restart DDEV to apply any changes:
```
```suggestion
- Restart DDEV to apply changes:
```

rfay commented on README.md:
````diff
@@ -2,77 +2,80 @@
 
-```yaml
-  - exec: printf "SHELL=/bin/bash\n* * * * * date |& tee -a /var/www/html/time.log\n* * * * * IS_DDEV_PROJECT=true /var/www/html/vendor/bin/typo3 scheduler:run -vv |& tee -a /var/www/html/scheduler-log.txt\n" | crontab
+A cron to add on to the example and then run the TYPO3 scheduler every minute might be:
````
Add words to explain where they would put this cron. They probably skimmed over above. Best to put in an extra file if possible.

tyler36 replied:
Update example sections

rfay reviewed (changes requested):
This is awesome. A few problems with copy-pasta, but it's great. 

The overall problem of how to manage custom configuration without breaking future `ddev get ddev/ddev-cron` and also making `ddev get --remove` work successfully has possibilities. In add-ons I've been using unmanaged files (non-#ddev-generated) like `docker-compose.redis-extra.yaml` for example as overrides. Not sure how easy that is to do here but I think you'll get it.

tyler36 commented:
Wow ... Not sure where my head was that day. Sorry for the low quality.

tyler36 committed:
Update README.md

Co-authored-by: Randy Fay <randy@randyfay.com>

tyler36 committed:
Update README.md

Co-authored-by: Randy Fay <randy@randyfay.com>

tyler36 committed:
Update README.md

Co-authored-by: Randy Fay <randy@randyfay.com>

tyler36 committed:
Use `COPY` command because it's preferred
https://www.geeksforgeeks.org/difference-between-the-copy-and-add-commands-in-a-dockerfile/

tyler36 committed:
Dynamically add `.cron` files

tyler36 committed:
Concatenate project cron jobs

tyler36 committed:
attempt to add cron jobs to current user

tyler36 commented:
Did some research and had a rethink.

- No default job is present OOTB
- Update now allows multiple `*.cron` jobs
- Examples updated to show specific steps.

tyler36 committed:
centralize health checks and example tests

tyler36 commented:
I suspect the "install from release" test are failing because they don't the content from this PR, no?

rfay commented:
Yes, you can either ignore the "install from release" tests or temporarily disable them.

rfay commented on README.md:
```diff
@@ -2,77 +2,124 @@
 ## Implementation
 
-The provided `web-build/Dockerfile.ddev-cron` and `web-build/cron.conf` configure the traditional cron daemon to run inside the web container.
+This extension does the follow:
```
```suggestion
This extension does the following:
```

rfay replied:
This one needs to be committed, should be "following"

rfay commented on README.md:
````diff
@@ -2,77 +2,124 @@
-```yaml
-  - exec: printf "SHELL=/bin/bash\n*/10 * * * * IS_DDEV_PROJECT=true DDEV_PHP_VERSION=8.0 /var/www/html/vendor/bin/drush cron -v |& tee -a /var/www/html/cron-log.txt\n" | crontab
+```cron
+* * * * * cd /var/www/html && IS_DDEV_PROJECT=true php artisan schedule:run >> /dev/null 2>&1
````
I wouldn't have thought this could have gotten installed in any way if IS_DDEV_PROJECT were not true in this context.

tyler36 replied:
Confirmed; tested on Laravel 10. 
Removed "IS_DDEV_PROJECT" and it still ran the command.

rfay replied:
The reason it worked on Laravel (but wouldn't work on Drupal) is that Laravel uses a .env at the project level, which is not protected by $IS_DDEV_PROJECT. Drupal/drush wouldn't load the settings.ddev.php for example. Thanks for testing and making sure these things worked despite my forgetting why that was there.

rfay commented:
Test with 
```
ddev get https://github.com/tyler36/ddev-cron/tarball/issue31_Move-to-Dockerfile-installation
```

rfay commented on README.md:
````diff
@@ -2,77 +2,124 @@
+  ddev get ddev/ddev-cron
+  ```
+
+- Customize the cron service, if required, by updating `./.ddev/web-build/cron.conf`
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
````
```suggestion
```

It just doesn't seem likely that anybody will need to do this.

tyler36 replied:
Yes, I _think_ it's gone now.

rfay commented on README.md:
````diff
@@ -2,77 +2,124 @@
+- Customize the cron service, if required, by updating `./.ddev/web-build/cron.conf`
+  - Remove `#ddev-generated` to prevent DDEV from overriding the file.
+- Add at least one `./ddev/web-build/*.cron` file. This will be automatically added to crontab on startup. See [Implementation](#implementation)
+- Restart DDEV to apply changes:
+
+  ```shell
+  ddev restart
+  ```
````
```suggestion
- `ddev restart` to apply changes.
```

rfay commented on README.md:
````diff
@@ -2,77 +2,124 @@
-    # `ddev exec crontab -l` will show you the current crontab configuration
-    - exec: printf "SHELL=/bin/bash\n* * * * * date >> /var/www/html/time.log\n" | crontab
-```
+### ./ddev/web-build/cron.conf
+
+This addon provides basic configuration for the cron service out of the box.
+
+To change the configuration for the cron service, update `./ddev/web-build/cron.conf`:
+
+- Remove `#ddev-generated` to prevent DDEV from overriding the file.
+- Make any required changes.
````
```suggestion
```

This is just a repeat, but it's unnecessary for any user we can imagine, true?

rfay replied:
This needs to be committed also, or discussed. You marked it resolved, but it wasn't committed. I don't think we need discussion of the cron.conf.

rfay reviewed (changes requested):
I tried this out and it looks great. Some suggestions along the way.

* I think it would be good to add info about `timezone: ` configuration here because people will get confused about what time things happen without that
* Mentioning `ddev exec crontab -l` may help people to see what's configured.

tyler36 committed:
Fix COPY command preventing multiple jobs

tyler36 commented:
OK ... merged the suggested changes.

Also fixed the issue with running multiple jobs so should be OK now.

rfay commented:
OH, now I remember why IS_DDEV_PROJECT is critical in the context of cron. Cron doesn't pass any environment variables along. So for any CMS type that has DDEV-generated settings, like Drupal or TYPO3 but probably not Laravel? the settings won't be executed without it. Sorry for forgetting this.

tyler36 commented:
Thanks. I thinks that everything now, or did I miss something?

rfay commented on README.md:
```diff
@@ -2,77 +2,124 @@
 - Installs and runs the cron service inside the web container
-- Adds a sample cron configuration that adds to a file every minute.
+- Adds an example job that writes out the current time.
 - Required DDEV v1.19.3 or higher.
```
I think we can remove this now. And IIRC the install.yaml checks and gives output anyway.

```suggestion
```

rfay reviewed (changes requested):
Several suggestions didn't get committed or discussed, please return to those. 

The one thing that might make it easier for people would be to provide the main examples as examples in the .ddev/web-build. Maybe the drupal and typo3 examples, for example, just like you do with the simple time log. 

Thanks for this!

tyler36 committed:
Fix typo

Co-authored-by: Randy Fay <randy@randyfay.com>

tyler36 committed:
Remove repeated section

Co-authored-by: Randy Fay <randy@randyfay.com>

tyler36 committed:
Remove requirement statement

Co-authored-by: Randy Fay <randy@randyfay.com>

tyler36 committed:
Use `crontab -l` in example because it is simpler

tyler36 commented:
> Mentioning ddev exec crontab -l may help people to see what's configured.

I replace `crontab -e` with `crontab -l` in the example.  `-e` asks about selecting an editor, `-l` simply outputs the contents

rfay commented:
Sorry, we're getting lost here, and the review page is getting too cluttered and I get lost in it now.

Here's what the README has right now:
https://github.com/ddev/ddev-cron/blob/642fe9ab2c94cfd4ad7f2ac0d5e3c40575954c58/README.md

This is one of the things I was trying to get rid of because I don't believe people will ever need to edit it:

<img width="953" alt="image" src="https://github.com/ddev/ddev-cron/assets/112444/154cd360-fa51-4c50-9afd-1ab3320675e9">

Other than thinking we shouldn't confuse people with that one, I think everything is fine to go. 

Sorry it's been so difficult to get through this, but I sure appreciate this good work.

tyler36 committed:
Remove customize instructions

tyler36 committed:
Remove extra line

tyler36 commented:
Appologies for the very messy work flow on this. The VSCode extension seemed inconsistent on this repo. And my fingers were too fat at other times. 

![ddev-cron-readme](https://github.com/ddev/ddev-cron/assets/7234392/07750a56-2046-464b-ad2c-09dd678be65d)

rfay reviewed (approved):
Thanks, looking great. Pull when you're ready. 

I don't think that github "Files changed" page was helping us much. It got all messy and you couldn't tell what was where.

tyler36 commented:
Thank you. 

We got there in the end.

tyler36 commented:
All tests from this PR passed after merge and `1.5.0` (which included this) was released.
